### PR TITLE
Convert to Stream only

### PR DESCRIPTION
**Pull request by digitalbuddha**
remove get api.  Some tests are broken but its ok since no one is using this yet. Would love feedback

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -23,7 +23,7 @@ buildscript {
     ]
 
     dependencies {
-        classpath 'com.android.tools.build:gradle:3.5.0-beta05'
```
Comment (yigit) [inline build.gradle]: any reason for this? or just living at the edge ?:)

File: store/src/main/java/com/nytimes/android/external/store3/base/AllPersister.kt [hunk 2]
```diff
@@ -26,7 +26,7 @@ interface AllPersister<Raw, Key> : Persister<Raw, Key>, DiskAllRead<Raw>, DiskAl
      * throw an exception
      */
     //    @Override
-    override suspend fun read(key: Key): Raw?
+    override suspend fun read(key: Key): Raw
```
Comment (yigit) [inline store/src/main/java/com/nytimes/android/external/store3/base/AllPersister.kt]: i wonder why you needed to change old Store code ?
do we lose the ability to implement `get` with this  new setup? (just `get` in old Store API, not pipeline)

File: store/src/main/java/com/nytimes/android/external/store3/pipeline/Pipeline.kt [hunk 3]
```diff
@@ -66,7 +66,7 @@ fun <Key, OldOutput, NewOutput> PipelineStore<Key, OldOutput>.withPersister(
 @ExperimentalCoroutinesApi
 @FlowPreview
 fun <Key, OldOutput, NewOutput> PipelineStore<Key, OldOutput>.withNonFlowPersister(
-    reader: suspend (Key) -> NewOutput?,
+    reader: suspend (Key) -> NewOutput,
```
Comment (yigit) [inline store/src/main/java/com/nytimes/android/external/store3/pipeline/Pipeline.kt]: i think this can stay nullable since it is just a single shot.

File: store/src/main/java/com/nytimes/android/external/store3/pipeline/PipelineCacheStore.kt [hunk 4]
```diff
@@ -14,20 +14,15 @@ internal class PipelineCacheStore<Key, Output>(
...
-                delegate.get(request)
-            },
             memoryPolicy = memoryPolicy ?: StoreDefaults.memoryPolicy
     )
 
     override fun stream(request: StoreRequest<Key>): Flow<Output> {
         @Suppress("RemoveExplicitTypeArguments")
         return flow<Output> {
-            if (!request.shouldSkipCache(CacheType.MEMORY)) {
+            if (request.shouldLoadFrom(CacheType.MEMORY)) {
```
Comment (yigit) [inline store/src/main/java/com/nytimes/android/external/store3/pipeline/PipelineCacheStore.kt]: liked positive method much better

File: store/src/main/java/com/nytimes/android/external/store3/pipeline/PipelineCacheStore.kt [hunk 5]
```diff
@@ -14,20 +14,15 @@ internal class PipelineCacheStore<Key, Output>(
...
                 val cached = memCache.getIfPresent(request.key)
-                cached?.let {
-                    emit(it)
-                }
+                cached?.let { emit(it) }
```
Comment (yigit) [inline store/src/main/java/com/nytimes/android/external/store3/pipeline/PipelineCacheStore.kt]: i guess we need a style guide :)

File: store/src/main/java/com/nytimes/android/external/store3/pipeline/PipelinePersister.kt [hunk 6]
```diff
@@ -2,71 +2,35 @@ package com.nytimes.android.external.store3.pipeline
...
-import kotlinx.coroutines.flow.flow
-import kotlinx.coroutines.flow.switchMap
+import kotlinx.coroutines.flow.*
 import kotlinx.coroutines.launch
-
 @FlowPreview
 class PipelinePersister<Key, Input, Output>(
         private val fetcher: PipelineStore<Key, Input>,
-        private val reader: (Key) -> Flow<Output?>,
+        private val reader: (Key) -> Flow<Output>,
```
Comment (yigit) [inline store/src/main/java/com/nytimes/android/external/store3/pipeline/PipelinePersister.kt]: as discussed in Slack, lets change this to be 
`StoreResponse`. Though I think the `Pipeline#stream` will be converted to `StoreResponse` not just this.

File: store/src/main/java/com/nytimes/android/external/store3/pipeline/PipelinePersister.kt [hunk 7]
```diff
@@ -2,71 +2,35 @@ package com.nytimes.android.external.store3.pipeline
...
-    @Suppress("UNCHECKED_CAST")
     override fun stream(request: StoreRequest<Key>): Flow<Output> {
-        if (request.shouldSkipCache(CacheType.DISK)) {
-            return fetcher.stream(request)
-                .switchMap {
-                    writer(request.key, it)
-                    reader(request.key)
-                }.castNonNull()
+        return if (request.shouldLoadFrom(CacheType.Persistent)) {
+            reader(request.key)
```
Comment (yigit) [inline store/src/main/java/com/nytimes/android/external/store3/pipeline/PipelinePersister.kt]: this is losing the ability to refresh while also fetching from network.
not sure why this change was necessary for the `stream` path.

File: store/src/main/java/com/nytimes/android/external/store3/pipeline/SimplePersisterAsFlowable.kt [hunk 8]
```diff
@@ -17,15 +17,17 @@ import kotlinx.coroutines.sync.withLock
...
     private val writer: suspend (Key, Input) -> Unit,
     private val delete: (suspend (Key) -> Unit)? = null
 ) {
     private val versionTracker = KeyTracker<Key>()
 
-    fun flowReader(key: Key): Flow<Output?> = flow {
-        versionTracker.keyFlow(key).collect {
-            emit(reader(key))
+    fun flowReader(key: Key): Flow<Output> {
+        return flow {
```
Comment (yigit) [inline store/src/main/java/com/nytimes/android/external/store3/pipeline/SimplePersisterAsFlowable.kt]: not liking the `=` shortcut? :'(

File: store/src/test/java/com/nytimes/android/external/store3/pipeline/PipelineStoreTest.kt [hunk 9]
```diff
@@ -57,116 +57,125 @@ class PipelineStoreTest {
...
-            key = key,
-            refresh = false
-        )
-    )
+    suspend fun <T, V> PipelineStore<T, V>.get(key: T) = stream(
+            StoreRequest.cached(
+                    key = key,
+                    refresh = false
+            )
+    ).singleOrNull()
```
Comment (fabioCollini) [inline store/src/test/java/com/nytimes/android/external/store3/pipeline/PipelineStoreTest.kt]: Looking at the documentation `singleOrNull` _Throws IllegalStateException for flow that contains more than one element_. I am not sure what's the best operator to use, maybe `first` with a try/catch or maybe `take(1). singleOrNull()`.

File: store/src/main/java/com/nytimes/android/external/store3/pipeline/PipelineStore.kt [hunk 10]
```diff
@@ -45,13 +32,15 @@ interface PipelineStore<Key, Output> {
 fun <Key, Output> PipelineStore<Key, Output>.open(): Store<Output, Key> {
     val self = this
     return object : Store<Output, Key> {
-        override suspend fun get(key: Key) = self.get(
-            StoreRequest.cached(key, refresh = false)
-        )!!
+        override suspend fun get(key: Key): Output {
+            return self.stream(
+                    StoreRequest.cached(key, refresh = false)
+            )!!.single()
```
Comment (fabioCollini) [inline store/src/main/java/com/nytimes/android/external/store3/pipeline/PipelineStore.kt]: do you need the `!!` here?

File: store/src/main/java/com/nytimes/android/external/store3/pipeline/PipelinePersister.kt [hunk 11]
```diff
@@ -1,100 +1,66 @@
...
-    return flow {
-        self.collect {
-            if (it != null) {
-                emit(it as T2)
-            }
-        }
-    }
-}
 
+private fun <T> Flow<T>.startWith2(networkFlow: Flow<T>): Flow<T> = flow {
```
Comment (fabioCollini) [inline store/src/main/java/com/nytimes/android/external/store3/pipeline/PipelinePersister.kt]: isn't there already an operator to do it? Something like `merge`?

Comment (fabioCollini) [reply store/src/main/java/com/nytimes/android/external/store3/pipeline/PipelinePersister.kt]: or `concat`

File: store/src/main/java/com/nytimes/android/external/store3/pipeline/PipelinePersister.kt [hunk 12]
```diff
@@ -1,100 +1,66 @@
...
-                    if (request.refresh) {
-                        // also request from backend
-                        it.sideCollect(fetcher.stream(request)) { response: Input ->
-                            response?.let { data: Input ->
-                                writer(request.key, data)
-                            }
+        return if (request.shouldLoadFrom(CacheType.Persistent)) {
+            return if (request.refresh) {
+                //we need to refresh let's emit the current value and then concat the fetcher/disk flow
+                reader(request.key).take(1).filter { it == null }.concatUpstreamFlow(request)
```
Comment (fabioCollini) [inline store/src/main/java/com/nytimes/android/external/store3/pipeline/PipelinePersister.kt]: is this `== null` or `!= null`?
